packages/ckeditor5-engine/src/model/differ.js

```diff
@@ -880,6 +931,26 @@ export default class Differ {
 					}
 				}
 			}
+
+			if ( inc.type === 'refresh' ) {
```

Comment:
> However, I am missing transformations for X vs "refresh". For example, if there is insert before refresh, refresh offset needs to be moved, etc. See other transformations, for example, insert vs insert.

That's probably because the `_isInInsertedElement()` that was added before most of the `refresh vs X` updates.

Reply:
Mmm... I don't understand your answer :D. Do you agree with me or do you say that what I mentioned is not needed ("refresh" being `old` change).

Reply:
I think that `X vs refresh` (I wrote the opposite) is handled by the `_isInInsertedElement` checks. It looks like handling `X vs Y` is done in two steps:

1.  when handing given change in `bufferOperation()`
2.  then afterward in `_handleChange()`

This migth be a reason why `X vs refresh` are not needed in `_handleChange()`.

OTOH, yes adding a test for `insert` before `refresh` (and similar) might be needed.

TL;DR - I agree with you (update the tests if needed)

Reply:
Writing from the top of my head but I think I remember how `_isInInsertedElement` works ;).

So, I think you are not right. Following scenario is probably not handled:

1.  Assume model: `<paragraph>Foo</paragraph><paragraph>Bar</paragraph>`.
2.  In one change block:
    1.  Refresh paragraph "Bar".
    2.  Insert paragraph at the beginning.

I think that in this scenario, the refresh change will not get updated, so the differ will return:

*   insert paragraph on index `0`,
*   refresh paragraph on index `1`.

So, in the end, a wrong paragraph will be refreshed.

Reply:
Thanks for the clarification :) ATM I'm still diving into the `Differ` internals. A test for such scenario(s) should be added.

Additionally, I'm thinking now about moving this to `DowncastDispatcher` anyway. But I don't want to rush into this without re-thinking this.

Reply:
Done on `DowncastDispatcher` level only.